### PR TITLE
Implement recording functionality with various updates

### PR DESCRIPTION
**Pull request by patelneel55**
Here's a list of all updates that include in this Record:

### CODE REVIEW

File: .jenkinsfile [hunk 1]
```diff
@@ -226,7 +226,7 @@ pipeline {
                                 sh '''
                                     echo "Running test suite on ${HL_OS}..."
 
-                                    lcov -c -i -d build-osx -o build-osx/coverage.base
+                                    # lcov -c -i -d build-osx -o build-osx/coverage.base
```
Comment (jmcker) [inline .jenkinsfile]: Uncomment this to get code coverage again

File: CMakeLists.txt [hunk 2]
```diff
@@ -175,8 +177,17 @@ if (NOT HL_BUILD_ONLY_AUDIO)
     endif ()
 endif ()
 
+# Required Sndfile packages
+find_package (SndFile REQUIRED)
```
Comment (jmcker) [inline CMakeLists.txt]: Wrap with NOT HL_ONLY_AUDIO

File: src/audio/CMakeLists.txt [hunk 3]
```diff
@@ -26,14 +26,16 @@ else ()
 endif ()
 
 # Add hlaudio library
-include_directories (${EXTRA_INCLUDES})
+# TODO: Remove ${INCLUDE_DIRS} once recording debug is complete
+include_directories (${EXTRA_INCLUDES} ${INCLUDE_DIRS})
```
Comment (jmcker) [inline src/audio/CMakeLists.txt]: ?

File: src/audio/CMakeLists.txt [hunk 4]
```diff
@@ -26,14 +26,16 @@ else ()
...
 add_library_target (${HL_LIBRARY_NAME} "${AUDIO_SRC_FILES};${EXTRA_INCLUDES}")
 
 # Force portaudio to build first
 add_dependencies (${HL_LIBRARY_NAME} portaudio portaudio_static)
 
 # Link external libraries to new generated library
-target_link_libraries (${HL_LIBRARY_NAME} ${AUDIO_LIBS})
+# TODO: Remove ${HL_LIBRARIES} once recording debug is completed
```
Comment (jmcker) [inline src/audio/CMakeLists.txt]: ?

File: src/ui/gui/qml/ButtonPanel.qml [hunk 5]
```diff
@@ -421,7 +451,7 @@ Rectangle {
             id: saveDialog
             objectName: "saveDialog"
             fileMode: FileDialog.SaveFile
-            nameFilters: saveDialog.nameFilters
+            nameFilters: ["WAVE Sound (*.wav)", "FLAC (*.flac)", "Core Audio Format (*.caf)", "Audio Interchange File Format (*.aiff)", "All files (*)"]
```
Comment (jmcker) [inline src/ui/gui/qml/ButtonPanel.qml]: Add RAW option here

File: src/control/Export.cpp [hunk 6]
```diff
@@ -17,32 +21,104 @@ Export::Export(std::string targetFile)
...
+    // Initialize libsndfile info.
+    SF_INFO sfinfo_out;
+    sfinfo_out.samplerate = SAMPLE_RATE;
+    sfinfo_out.channels = NUM_CHANNELS;
+    sfinfo_out.format = SF_FORMAT_FLOAT;
+    float* buffer = new float[512];
+
+    // Set libsndfile settings based on extension
+    // TODO: Compare against Encoding enum from HulaAudioSettings
+    if (!extension.compare("wav"))
```
Comment (jmcker) [inline src/control/Export.cpp]: Is this case sensitive?

File: src/control/Export.cpp [hunk 7]
```diff
@@ -17,32 +21,104 @@ Export::Export(std::string targetFile)
...
+/**
+ * Deletes all the files in the vector
+ */
+void Export::deleteTempFiles(std::vector<std::string> dirs)
+{
+    // loop throught all the files
+    for(std::string file : dirs)
+    {
+        // no good c++ function so we'll just use the C one
+        remove((char *)file.c_str());
```
[the remark this answers is not in the document]
Comment (patelneel55) [reply src/control/Export.cpp]: I was wrong...remove did work however when you discard recording restarts which in turn created another file....This is fixed in recordermerge I believe

**Pull request merged**